### PR TITLE
Add unit tests for Value and introduce module.

### PR DESCRIPTION
**Pull request by bgogul**
This PR cleans up the representation to create the following classes that depend on each other. 
 - A `Module` is a collection of `Block`s.
 - A `Block` is a collection of `Operation`s.
 - An `Operation` may have an optional `Module` that captures the semantics of the operation. This can be used for representing bodies of particles, functions, etc. 

This PR also simplifies `Value` to the barest minimum required. 

This PR deletes `block.h` and `operation.h` and moves their content to `module.h`.

### CODE REVIEW

File: src/ir/module.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  const Operator& op() const { return *op_; }
+  const Block* parent() const { return parent_; }
+  const NamedValueMap& inputs() const { return inputs_; }
+  const NamedAttributeMap& attributes() const { return attributes_; }
+
+  std::string ToString(SSANames& ssa_names) const;
+
+ private:
+  // The parent block if any to which this operation belongs to.
+  const Block* parent_;
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/module.h:56]: Yes, I expect that we will we will need this. e.g., to find the higher level operation under to which this belongs to.

File: src/ir/module.h [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  const DataDeclCollection& inputs() const { return inputs_; }
+  const DataDeclCollection& outputs() const { return outputs_; }
+  const NamedValueMap& results() const { return results_; }
+  const Module* module() const { return module_; }
+
+  friend class BlockBuilder;
+
+ private:
+  // Module to which this belongs to.
+  const Module* module_;
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/module.h:91]: Yes, I expect we will need this.

**Comment by bgogul** [comment]
@markww addressed/replied to your comments. PTAL.

**Review by markww — APPROVED** [review]
LGTM

**Review by bgogul — COMMENTED** [review]
Forgot to hit submit review for my replies to your comments.

**Pull request merged**